File: sdks/ruby/lib/novabilling/types/mrr_breakdown_response.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class MrrBreakdownResponse < Internal::Types::Model
      field :total_mrr, -> { Integer }, optional: false, nullable: false, api_name: "totalMrr"
      field :new_mrr, -> { Integer }, optional: false, nullable: false, api_name: "newMrr"
      field :expansion_mrr, -> { Integer }, optional: false, nullable: false, api_name: "expansionMrr"
      field :contraction_mrr, -> { Integer }, optional: false, nullable: false, api_name: "contractionMrr"
      field :churn_mrr, -> { Integer }, optional: false, nullable: false, api_name: "churnMrr"
      field :net_new_mrr, -> { Integer }, optional: false, nullable: false, api_name: "netNewMrr"
      field :by_plan, -> { Internal::Types::Array[Novabilling::Types::MrrPlanBreakdown] }, optional: false, nullable: false, api_name: "byPlan"
    end
  end
end

File: sdks/csharp/src/NovaBilling/Types/MrrBreakdownResponse.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record MrrBreakdownResponse : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("totalMrr")]
    public required double TotalMrr { get; set; }

    [JsonPropertyName("newMrr")]
    public required double NewMrr { get; set; }

    [JsonPropertyName("expansionMrr")]
    public required double ExpansionMrr { get; set; }

    [JsonPropertyName("contractionMrr")]
    public required double ContractionMrr { get; set; }

    [JsonPropertyName("churnMrr")]
    public required double ChurnMrr { get; set; }

    [JsonPropertyName("netNewMrr")]
    public required double NetNewMrr { get; set; }

    [JsonPropertyName("byPlan")]
    public IEnumerable<MrrPlanBreakdown> ByPlan { get; set; } = new List<MrrPlanBreakdown>();

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
